### PR TITLE
Adding the otterspace bagdes strategy

### PR DESCRIPTION
**Pull request by rahulrumalla**
cc @grp06 @firatkaratas @stefan-girlich please review this PR

--

Some raw relevant notes from snapshot team

> Hi @rsquare! The voting strategies resolve voting power for multiple addresses at once, if you want make voting power based on NFT attribute(s) you need to way to query those attributes for multiple addresses in an efficient way. Doing this directly with a node and IPFS gateway take too much requests and doesnt scale. What I recommend is to create a subgraph that store these informations (which address hold which NFT attribute) and from this you can make a custom voting strategy that resolve voting power in just few requests.
> 
> Subgraph are on Optimism we use subgraph there for others things like delegation. But if you like you can use your own API too or another solution, the only requirement is that you can calculate the voting power at a specific time or block number, having an API with just the voting power is not enough

--

Goal: Create a snapshot strategy for DAOs so that they can create governance proposals where badge holders can vote in a 1 badge = 1 vote manner. Also provide option to speficy varying degrees of influence across various sub-daos or different teams/guilds i.e., specified via a badge spec. 

Solution: A custom snapshot strategy that will talk to our subgraph API

--

Basically, in snapshot, when a vote is cast it is associated with "weights" to compute a "score" for that vote. 

So for DAOs for exampl, we can configure all core contribs have a weight of 100 vs Members havign a weight of 1. When no weights are specified, I made the logic to be that all token are then equal with a weight of 1. 

When setting up the snapshot prroposal, a user is presented an option to select a strategy. Upon selecting, u have to provide a "configuration", which is what is in the example file.
image

### CODE REVIEW

File: src/strategies/otterspace-badges/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+* A RAFT is an NFT that represents a community(or a DAO) within Otterspace.
+* In order to mint a badge, it is defined with a badge spec, which is essentially dscribed as ERC721 metadata standard. A RAFT has several badge specs
+* Each BADGE is a non-transferable token built with [ERC4973]("https://github.com/otterspace-xyz/ERC4973") spec and maximally backwards compatible with ERC721. Each badge spec may have several badges associated which indicate the badges minted for that spec that is associated to a raft
+
+The parameters when using this strategy must specify
+* A RAFT tokenId
+* The RAFT's contract address
+* An array of weights associated to the badge specs
+
+If no specs are specifices, all badges under the RAFT are considered equal with the weight of 1
```
Comment (firatkaratas) [inline src/strategies/otterspace-badges/README.md]:
```suggestion
If no specs are specified, all badges under the RAFT are considered equal with the weight of 1
```

File: src/strategies/otterspace-badges/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# Otterspace Badges
+* A RAFT is an NFT that represents a community(or a DAO) within Otterspace.
+* In order to mint a badge, it is defined with a badge spec, which is essentially dscribed as ERC721 metadata standard. A RAFT has several badge specs
```
Comment (firatkaratas) [inline src/strategies/otterspace-badges/README.md]:
```suggestion
* In order to mint a badge, it is defined with a badge spec, which is essentially described as ERC721 metadata standard. A RAFT has several badge specs
```

File: src/strategies/otterspace-badges/README.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+* Each BADGE is a non-transferable token built with [ERC4973]("https://github.com/otterspace-xyz/ERC4973") spec and maximally backwards compatible with ERC721. Each badge spec may have several badges associated which indicate the badges minted for that spec that is associated to a raft
+
+The parameters when using this strategy must specify
+* A RAFT tokenId
+* The RAFT's contract address
+* An array of weights associated to the badge specs
+
+If no specs are specifices, all badges under the RAFT are considered equal with the weight of 1
+
+Here is an example of it's usage:
```
Comment (stefan-girlich) [inline src/strategies/otterspace-badges/README.md]: _"its"_

File: src/strategies/otterspace-badges/README.md (hunk 1 above)
Comment (stefan-girlich) [inline src/strategies/otterspace-badges/README.md]: _"specified"_

File: src/strategies/otterspace-badges/README.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
+# Otterspace Badges
+* A RAFT is an NFT that represents a community(or a DAO) within Otterspace.
```
Comment (stefan-girlich) [inline src/strategies/otterspace-badges/README.md]: missing space after _"community"_

File: src/strategies/otterspace-badges/README.md (hunk 2 above)
Comment (stefan-girlich) [inline src/strategies/otterspace-badges/README.md]: - _"In order to mind [...] it is defined [...] which is described [...]"_ - sentence seems a bit bulky overall. should we get rid of the passive forms and explain who creates the specs and who mints?
- _"described"_

File: src/strategies/otterspace-badges/README.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+* In order to mint a badge, it is defined with a badge spec, which is essentially dscribed as ERC721 metadata standard. A RAFT has several badge specs
+* Each BADGE is a non-transferable token built with [ERC4973]("https://github.com/otterspace-xyz/ERC4973") spec and maximally backwards compatible with ERC721. Each badge spec may have several badges associated which indicate the badges minted for that spec that is associated to a raft
+
+The parameters when using this strategy must specify
+* A RAFT tokenId
```
Comment (stefan-girlich) [inline src/strategies/otterspace-badges/README.md]: _"token ID"_?

File: src/strategies/otterspace-badges/index.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
+import { subgraphRequest } from '../../utils';
+import examplesFile from './examples.json';
+
+export const author = 'otterspace';
+export const version = '1.0.0';
+export const examples = examplesFile;
```
Comment (stefan-girlich) [inline src/strategies/otterspace-badges/index.ts]: unused

File: src/strategies/otterspace-badges/index.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+const OTTERSPACE_SUBGRAPH_API_URL = {
```
Comment (stefan-girlich) [inline src/strategies/otterspace-badges/index.ts]: `OTTERSPACE_SUBGRAPH_API_URLS` or `SUBGRAPH_URLS_BY_CHAIN_ID`?

File: src/strategies/otterspace-badges/index.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+export async function strategy(
+  space,
+  network,
+  provider,
+  addresses,
+  options,
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  snapshot
+) {
+  // Get all the badges for a raft
```
Comment (stefan-girlich) [inline src/strategies/otterspace-badges/index.ts]: - comment _"Get all the badges for a raft"_ could be a file-level function `fetchBadgesForRaft(network: string, raftAddress: string, raftTokenId: string)`
- mutating a file-level constant is unusual; we could instead use a function, e.g.  `const params = buildSubgraphQueryParams(options.raftAddress, options.raftTokenId)`
- who ensures `network` is a valid key for `OTTERSPACE_SUBGRAPH_API_URL`? should we handle errors explicitly?
- same question for potential failure of `subgraphRequest()`; assumption: letting thrown error uncaught is fine

File: src/strategies/otterspace-badges/index.ts [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+  params.badges.__args.where.spec_.raft = `${options.raftAddress}:${options.raftTokenId}`;
+  const getBadgesResponse = await subgraphRequest(
+    OTTERSPACE_SUBGRAPH_API_URL[network],
+    params
+  );
+
+  // If a badge's spec is listed in the options, set the associated weight
```
Comment (stefan-girlich) [inline src/strategies/otterspace-badges/index.ts]: - extract logic from this line downwards to function `applyBadgeWeights`
- would remove inline comments since they only describe the code lines below them - which are very readable by themselves already

File: src/strategies/otterspace-badges/index.ts [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+  const badgeWeights = {};
+  if (getBadgesResponse && getBadgesResponse.badges) {
```
Comment (stefan-girlich) [inline src/strategies/otterspace-badges/index.ts]: we could save a level of indentation and return early, e.g. `if(!getBadgesResponse?.badges) return badgeWeights`

File: src/strategies/otterspace-badges/index.ts [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    getBadgesResponse.badges.forEach((badge: any) => {
```
Comment (stefan-girlich) [inline src/strategies/otterspace-badges/index.ts]: - extract interfaces for response types, maybe subgraph offers generated TypeScript code?

File: src/strategies/otterspace-badges/index.ts [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+      // set the weight of a badge as 0
+      let badgeWeight = badgeWeights[badgeAddress];
+      if (!badgeWeight) {
+        badgeWeight = 0;
+      } else {
+        return;
+      }
+
+      if (options.specs && options.specs.length > 0) {
```
Comment (stefan-girlich) [inline src/strategies/otterspace-badges/index.ts]: could be checked/sanitized once, outside of this loop, e.g. `const badgeSpecs = options.specs?.length ? options.specs : []`

File: src/strategies/otterspace-badges/index.ts [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+  );
+
+  // If a badge's spec is listed in the options, set the associated weight
+  const badgeWeights = {};
+  if (getBadgesResponse && getBadgesResponse.badges) {
+    getBadgesResponse.badges.forEach((badge: any) => {
+      const badgeAddress = badge.owner.toLowerCase();
+
+      // set the weight of a badge as 0
+      let badgeWeight = badgeWeights[badgeAddress];
```
Comment (stefan-girlich) [inline src/strategies/otterspace-badges/index.ts]: - simplify: `if(badgeWeights[badgeAddress]) return`

File: src/strategies/otterspace-badges/index.ts [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+      if (!badgeWeight) {
+        badgeWeight = 0;
+      } else {
+        return;
+      }
+
+      if (options.specs && options.specs.length > 0) {
+        // for the specs defined in config, set the associated weight
+        const specConfig = options.specs.find(
+          (s: any) => s.id === badge.spec.id
```
Comment (stefan-girlich) [inline src/strategies/otterspace-badges/index.ts]: - use explicit type/interface
- `s` => `spec`

File: src/strategies/otterspace-badges/index.ts [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+        );
+        badgeWeight = specConfig !== undefined ? specConfig.weight : 0;
```
Comment (stefan-girlich) [inline src/strategies/otterspace-badges/index.ts]: `badgeWeight = specConfig ? specConfig.weight : 0;`

File: src/strategies/otterspace-badges/index.ts [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+        // are treated as equally weighted
+        badgeWeight = 1;
+      }
+      badgeWeights[badgeAddress] = badgeWeight;
+    });
+  }
+
+  // For all the addresses that vote with a badge, perform a lookup to fetch their associated weight
+  return Object.fromEntries(
+    addresses.map((address: any) => [
```
Comment (stefan-girlich) [inline src/strategies/otterspace-badges/index.ts]: ´any´ => `string`

File: src/strategies/otterspace-badges/index.ts [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+      address,
+      badgeWeights[address.toLowerCase()] || 0
```
Comment (stefan-girlich) [inline src/strategies/otterspace-badges/index.ts:93]: case "0" already seems covered in line 54

**Review by stefan-girlich — CHANGES_REQUESTED** [review]

File: src/strategies/otterspace-badges/index.ts (hunk 6 above)
Comment (rahulrumalla) [reply src/strategies/otterspace-badges/index.ts]: It's used in testsI blv.

Comment (rahulrumalla) [reply src/strategies/otterspace-badges/index.ts]: ah u mean examplesFile?

File: src/strategies/otterspace-badges/index.ts (hunk 8 above)
Comment (rahulrumalla) [reply src/strategies/otterspace-badges/index.ts]: I don't think we need to handle errors explicitly. If there's an error, snapshot's UI will probably just show that it errored out. Not sure yet to how debug such situations as we can't do end-to-end testing locally.

File: src/strategies/otterspace-badges/index.ts (hunk 11 above)
Comment (rahulrumalla) [reply src/strategies/otterspace-badges/index.ts]: They unfortunately don't

File: src/strategies/otterspace-badges/index.ts (hunk 6 above)
Comment (stefan-girlich) [reply src/strategies/otterspace-badges/index.ts]: ah sorry, i forgot that it's exported and may be used elsewhere. but then i'd wonder why it's imported here and not at the site of usage (e.g. in the tests)

**Review by stefan-girlich — APPROVED** [review]
👍

**Comment by rahulrumalla** [comment]
Will issue a new PR directly to snapshot
